### PR TITLE
Renames ComposeViewFactoryRoot to CompositionRoot and decouples the implementation.

### PR DESCRIPTION
**Pull request by zach-klippenstein**
The root is now applied via a `ViewRegistry` wrapper that wraps individual factories
to apply the root, instead of having this logic hard-coded inside `ComposeViewFactory`.

This approach is more flexible in general (could be used to do other tricks), and decouples
the rooting feature from the rest of the code.

### CODE REVIEW

File: core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ViewRegistries.kt [hunk 1]
```diff
@@ -40,3 +41,25 @@ import com.squareup.workflow.ui.ViewRegistry
...
+  ): ViewFactory<RenderingT> {
+    val transformedFactory = transform(this@mapFactories.getFactoryFor(renderingType))
+    check(transformedFactory.type == renderingType) {
+      "Expected transform to return a ViewFactory that is compatible with $renderingType, " +
+          "but got one with type ${transformedFactory.type}"
+    }
+    @Suppress("UNCHECKED_CAST")
+    return transformedFactory as ViewFactory<RenderingT>
+  }
+}
```
Comment (zach-klippenstein) [inline core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ViewRegistries.kt:65]: This is the key to the decoupled implementation. This technique can be used to do other stuff too, like e.g. apply a modifier to each ViewFactory's view.

**Comment by zach-klippenstein** [comment]
Failed test looks like a weird github issue – same shard appears twice, passed the second time.

File: core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ViewRegistries.kt (hunk 1 above)
Comment (rjrjr) [reply core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ViewRegistries.kt:65]: I think I've already had a few use cases for this, should we move it to the main repo?

Comment (zach-klippenstein) [reply core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ViewRegistries.kt:65]: Sure!

**Pull request merged**